go/vt/schemadiff/column.go

```diff
@@ -79,9 +79,90 @@ func NewColumnDefinitionEntity(c *sqlparser.ColumnDefinition) *ColumnDefinitionE
 
 // ColumnDiff compares this table statement with another table statement, and sees what it takes to
 // change this table to look like the other table.
-// It returns an AlterTable statement if changes are found, or nil if not.
-// the other table may be of different name; its name is ignored.
-func (c *ColumnDefinitionEntity) ColumnDiff(other *ColumnDefinitionEntity, _ *DiffHints) *ModifyColumnDiff {
+// It returns an ModifyColumnDiff statement if changes are found, or nil if not.
+// The function also requires the charset/collate on the source & target tables. This is because the column's
+// charset & collation, if undefined, are really defined by the table's charset & collation.
+//
+//	Anecdotally, in CreateTableEntity.normalize() we actually actively strip away the charset/collate properties
+//	from the column definition, to get a cleaner table definition.
+//
+// Things get complicated when we consider hints.TableCharsetCollateStrategy. Consider this test case:
+//
+//	from: "create table t (a varchar(64)) default charset=latin1",
+//	to:   "create table t (a varchar(64) CHARACTER SET latin1 COLLATE latin1_bin)",
+//
+// In both cases, the column is really a latin1. But the tables themselves have different collations.
+// In TableCharsetCollateStrict, the situation is handled correctly by converting the table charset as well as
+// pinning the column's charset.
+// But in TableCharsetCollateIgnoreEmpty or in TableCharsetCollateIgnoreAlways, schemadiff is not supposed to
+// change the table's charset. In which case it does need to identify the fact that the two columns are really of
+// the same charset, and not report a diff.
+func (c *ColumnDefinitionEntity) ColumnDiff(
+	other *ColumnDefinitionEntity,
+	hints *DiffHints,
+	t1cc *charsetCollate,
+	t2cc *charsetCollate,
+) *ModifyColumnDiff {
+	// We need to understand whether the table's charset is going to be ignored. If so, we need to denormalize
+	// the columns' charset/collate properties for the sake of this comparison.
+	// Denormalization means that if the column does not have charset/collate defined, then it must borrow those
+	// propeties from the table.
+	denormalizeCharsetCollate := false
+	if c.IsTextual() || other.IsTextual() {
+		switch hints.TableCharsetCollateStrategy {
+		case TableCharsetCollateStrict:
+			// No need to do anything
```

Comment:
I don't think this is correct. Also in this case we can get an empty / column charset due to normalization if you change the table level charset and we shouldn't then generate a diff. 

This is actually visible in a broken test that was added.